backend/util/logger.py

```diff
@@ -0,0 +1,107 @@
+import logging
+import logging.handlers
+import queue
+import sys
+import traceback
+
+import config
+
+from typing import Union
+from multiprocessing import Queue
+from util.logFormatter import MultiLineFormatter
+
+
+class MultiprocessingLogger:
+
+    _logging_queue = None
+    _configured_loggers = {}
+
+    _processing_logger = None
+
+    @classmethod
+    def initialize_logger(cls, logging_queue: Union[Queue, None] = None) -> None:
+        """ Initialize the logger.
+
+        Must be called in the base process that start the other processes without any argument.
+        In other process it must be called with a dedicated logging queue."""
+        if logging_queue is None:
+            logging_queue = Queue()
+
+        cls._logging_queue = logging_queue
+
+    @classmethod
+    def get_logging_queue(cls) -> Union[Queue, None]:
+        return cls._logging_queue
+
+    @classmethod
+    def get_logger(cls, logger_name: str) -> logging.Logger:
+        # We only need to configure once, multiple calls to logging.getLogger(<logger_name>)
+        # will return the same configured instance
+        if logger_name in cls._configured_loggers:
+            return logging.getLogger(logger_name)
+
+        logger = logging.getLogger(logger_name)
+        logger.addHandler(logging.handlers.QueueHandler(cls._logging_queue))
+        logger.setLevel(config.microlabConfig.logLevel)
+
+        cls._configured_loggers[logger_name] = True
+
+        return logger
+
+    @classmethod
+    def cleanup_logging(cls):
+        # Without the retries check added here I would occasionally see a hang here
+        # on cleanup
+        retries = 0
+        while not cls._logging_queue.empty() and retries < 5:
+            cls.process_logs()
+            retries += 1
```

Comment:
Actually just changing `get_logger` to `_get_processing_logger` in main is a bad way to fix this, since in the main process any other file that logs would also need to change, and as both the main process and some subprocesses could call functions in the same file that log, (config.py for instance) we can't do that generally. 

Maybe just if MultiprocessingLogger has to create a queue, we can set a static variable like `_isMainProcess=true`, and then change `get_logger`  to instead call `_get_processing_logger` if it is?

Reply:
Good catch on all of this, that was the right approach for sure. My brain is still a bit melty from all of the crunch and you can see in the commit history it took me a minute to work through and get to the same place (apologies I can make a new branch sans those commits if we want a cleaner history).